server/aesdsocket.c

```diff
@@ -105,6 +152,187 @@ int demonize() {
     return 0;
 }
 
+void exit_cleaner() {
+
+    pthread_mutex_destroy(&mutex);
+
+    syslog(LOG_DEBUG, "Caught Signal, Exiting\n");
+        
+    if(sfd > 0)
+        close(sfd);
+       
+    if(fd > 0)
+        close(fd);
+    
+    
+    closelog();
+    
+    unlink("/var/tmp/aesdsocketdata");
+
+}
+// if true loop until head == NULL
+void check_thread_join(bool val) {
+
+   do{ 
+        current = head;
+        previous = head;
+        while(current != NULL) {
+            //Updating head if first node is complete
+            if (current->client_meta.t_complete == true) {
```

Comment:
@sachininja when exiting the application you need to join every thread.  Otherwise if one is running and hasn't yet completed by the time you reach this point you won't pthread_join and free associated memory.  This might be what valgrind was complaining about.